### PR TITLE
Issue 5980: Proxying requests from SafeBrowsingNetworkContext

### PR DESCRIPTION
**Pull request by jumde**
Fix https://github.com/brave/brave-browser/issues/5980

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [ ] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- [ ] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [ ] Ran `git rebase master` (if needed).
- [ ] Ran `git rebase -i` to squash commits (if needed).
- [ ] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [ ] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protection-Mode
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Test Plan:

1. Open Brave Browser
2. Verify that `Safe Browsing` store files are correctly downloaded to the `<data-dir>/Safe Browsing` directory
3. Navigate to https://testsafebrowsing.appspot.com
4. Try downloading a file and verify that the connection to sb-ssl.brave.com is initiated.

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

**Comment by bsclifton** [comment]
Added to 0.69.x with https://github.com/brave/brave-core/pull/3433
Added to 0.70.x with https://github.com/brave/brave-core/pull/3408
Added to 0.71.x with https://github.com/brave/brave-core/pull/3452

File: patches/components-safe_browsing-browser-safe_browsing_network_context.cc.patch [hunk 1]
```diff
@@ -0,0 +1,13 @@
+diff --git a/components/safe_browsing/browser/safe_browsing_network_context.cc b/components/safe_browsing/browser/safe_browsing_network_context.cc
+index ddca8cdc9d5eab41463ad9f1fc3893c0a8d23d6f..e97846dd8aa69b64e01d56c8803bb8652bc9d13c 100644
+--- a/components/safe_browsing/browser/safe_browsing_network_context.cc
++++ b/components/safe_browsing/browser/safe_browsing_network_context.cc
+@@ -80,7 +80,7 @@ class SafeBrowsingNetworkContext::SharedURLLoaderFactory
+                                 traffic_annotation) override {
+     DCHECK(content::BrowserThread::CurrentlyOn(content::BrowserThread::UI));
+     GetURLLoaderFactory()->CreateLoaderAndStart(
+-        std::move(loader), routing_id, request_id, options, request,
++        std::move(loader), routing_id, request_id, options, brave::OnBeforeSystemRequest(request),
```
Comment (simonhong) [inline patches/components-safe_browsing-browser-safe_browsing_network_context.cc.patch]: I think calling chrome(brave) layer api from components is layer violation.

Comment (jumde) [reply patches/components-safe_browsing-browser-safe_browsing_network_context.cc.patch]: Oh ok, i'm working on a fix for extension proxies: https://github.com/brave/brave-browser/issues/6331 - Will try to clean this up in the same patch.

Comment (bridiver) [reply patches/components-safe_browsing-browser-safe_browsing_network_context.cc.patch]: my suggestion to @yrliou is to create a singleton class in `brave/net` and then register callbacks on it to fix the layering violation

Comment (bridiver) [reply patches/components-safe_browsing-browser-safe_browsing_network_context.cc.patch]: cc @iefremov

**Pull request merged**